packages/core/admin/server/services/api-token.js

```diff
@@ -47,28 +125,60 @@ const hash = (accessKey) => {
 };
 
 /**
+ * Create a token and its permissions
+ *
  * @param {Object} attributes
  * @param {TokenType} attributes.type
  * @param {string} attributes.name
- * @param {string} [attributes.description]
+ * @param {string[]} attributes.permissions
+ * @param {string} attributes.description
  *
  * @returns {Promise<ApiToken>}
  */
 const create = async (attributes) => {
   const accessKey = crypto.randomBytes(128).toString('hex');
 
+  assertCustomTokenPermissionsValidity(attributes);
+
+  // Create the token
   const apiToken = await strapi.query('admin::api-token').create({
     select: SELECT_FIELDS,
+    populate: POPULATE_FIELDS,
     data: {
-      ...attributes,
+      ...omit('permissions', attributes),
       accessKey: hash(accessKey),
     },
   });
 
-  return {
-    ...apiToken,
-    accessKey,
-  };
+  const result = { ...apiToken, accessKey };
+
+  // If this is a custom type token, create and the related permissions
+  if (attributes.type === constants.API_TOKEN_TYPE.CUSTOM) {
+    // TODO: createMany doesn't seem to create relation properly, implement a better way rather than a ton of queries
+    // const permissionsCount = await strapi.query('admin::token-permission').createMany({
+    //   populate: POPULATE_FIELDS,
+    //   data: attributes.permissions.map(action => ({ action, token: apiToken })),
+    // });
+    await Promise.all(
+      attributes.permissions.map((action) =>
+        strapi.query('admin::token-permission').create({
+          data: { action, token: apiToken },
+        })
+      )
+    );
+
+    const currentPermissions = await strapi.entityService.load(
+      'admin::api-token',
+      apiToken,
+      'permissions'
+    );
+
+    if (currentPermissions) {
+      Object.assign(result, { permissions: map('action', currentPermissions) });
```

Comment:
Same as my previous comment, maybe we shouldn't flatten the permission objects when dealing with actual ApiToken entities